x-pack/plugins/serverless_search/public/application/components/connectors/edit_service_type.tsx

```diff
@@ -62,24 +64,125 @@ export const EditServiceType: React.FC<EditServiceTypeProps> = ({ connector, isD
     },
   });
 
+  const getInitialOptions = () => {
+    return allConnectors.map((conn, key) => {
+      const _append: JSX.Element[] = [];
+      if (conn.isTechPreview) {
+        _append.push(
+          <EuiBadge key={key + '-preview'} iconType="beaker" color="hollow">
+            {i18n.translate(
+              'xpack.serverlessSearch.connectors.chooseConnectorSelectable.thechPreviewBadgeLabel',
+              { defaultMessage: 'Tech preview' }
+            )}
+          </EuiBadge>
+        );
+      }
+      if (conn.isBeta) {
+        _append.push(
+          <EuiBadge key={key + '-beta'} iconType={'beta'} color="hollow">
+            {i18n.translate(
+              'xpack.serverlessSearch.connectors.chooseConnectorSelectable.BetaBadgeLabel',
+              {
+                defaultMessage: 'Beta',
+              }
+            )}
+          </EuiBadge>
+        );
+      }
+      return {
+        _append,
+        _prepend: <EuiIcon size="l" type={conn.iconPath} />,
+        key: key.toString(),
+        label: conn.name,
+        serviceType: conn.serviceType,
+      };
+    });
+  };
+
+  const initialOptions = getInitialOptions();
+  const { euiTheme } = useEuiTheme();
+  const [selectedOption, setSelectedOption] = useState<Array<EuiComboBoxOptionOption<OptionData>>>(
+    []
+  );
+  const [selectedConnector, setSelectedConnector] = useState<Connector | null>(null);
+  const renderOption = (
+    option: EuiComboBoxOptionOption<OptionData>,
+    searchValue: string,
+    contentClassName: string
+  ) => {
+    const { _append, key, label, _prepend, serviceType } =
+      option as EuiComboBoxOptionOption<OptionData> & {
+        _append: JSX.Element[];
```

Comment:
nitpick: `React.ReactNode` is more flexible then `JSX.Element` in my experience

Reply:
Good to know. Changed.